pkg/intern/intern.go

```diff
@@ -0,0 +1,41 @@
+package intern
+
+import "sync"
+
+// A Value pointer is the handle to an underlying comparable value.
+type Value struct {
+	cmpVal interface{}
+}
+
+// Get returns the comparable value passed to the Get func that returned v.
+func (v *Value) Get() interface{} { return v.cmpVal }
+
+// Our pool of interned values and a lock to serialize access.
+var (
+	mu  sync.Mutex
+	val = map[interface{}]*Value{}
+)
+
+// Get returns a pointer representing the comparable value cmpVal.
+//
+// The returned pointer will be the same for Get(v) and Get(v2)
+// if and only if v == v2, and can be used as a map key.
+//
+// Note that Get returns a *Value so we only return one word of data
+// to the caller, despite potentially storing a large amount of data
+// within the Value itself.
+func Get(cmpVal interface{}) *Value {
```

Comment:
i believe empty interface will alloc. we should either make this generic or specifically for strings.

Reply:
Interesting.  It think I'll pause here and try the proto writer for the blocklist instead and see what those look like.
```
goos: linux
goarch: amd64
pkg: github.com/grafana/tempo/tempodb/backend
cpu: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
                  │   old.txt    │               new.txt               │
                  │    sec/op    │    sec/op     vs base               │
IndexMarshal-16     161.9µ ± ∞ ¹   164.9µ ± ∞ ¹        ~ (p=0.056 n=5)
IndexUnmarshal-16   37.53µ ± ∞ ¹   53.70µ ± ∞ ¹  +43.09% (p=0.008 n=5)
geomean             77.94µ         94.09µ        +20.71%
¹ need >= 6 samples for confidence interval at level 0.95

                  │    old.txt    │               new.txt                │
                  │     B/op      │     B/op       vs base               │
IndexMarshal-16     1.033Mi ± ∞ ¹   1.033Mi ± ∞ ¹        ~ (p=0.690 n=5)
IndexUnmarshal-16   45.39Ki ± ∞ ¹   54.20Ki ± ∞ ¹  +19.41% (p=0.008 n=5)
geomean             219.2Ki         239.5Ki         +9.28%
¹ need >= 6 samples for confidence interval at level 0.95

                  │   old.txt   │               new.txt                │
                  │  allocs/op  │  allocs/op    vs base                │
IndexMarshal-16     50.00 ± ∞ ¹    50.00 ± ∞ ¹         ~ (p=0.278 n=5)
IndexUnmarshal-16   58.00 ± ∞ ¹   190.00 ± ∞ ¹  +227.59% (p=0.008 n=5)
geomean             53.85          97.47         +80.99%
¹ need >= 6 samples for confidence interval at level 0.95
```

Reply:
Comparing `interface{}` to `string` here.
```
❯ benchstat new.txt next.txt
goos: linux
goarch: amd64
pkg: github.com/grafana/tempo/tempodb/backend
cpu: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
                  │   new.txt   │             next.txt              │
                  │   sec/op    │   sec/op     vs base              │
IndexMarshal-16     159.4µ ± 8%   160.8µ ± 8%       ~ (p=0.959 n=8)
IndexUnmarshal-16   65.92µ ± 2%   67.20µ ± 2%  +1.93% (p=0.028 n=8)
geomean             102.5µ        103.9µ       +1.41%

                  │   new.txt    │              next.txt              │
                  │     B/op     │     B/op      vs base              │
IndexMarshal-16     1.033Mi ± 0%   1.033Mi ± 0%       ~ (p=1.000 n=8)
IndexUnmarshal-16   55.44Ki ± 0%   55.25Ki ± 0%  -0.34% (p=0.000 n=8)
geomean             242.1Ki        241.8Ki       -0.15%

                  │  new.txt   │             next.txt             │
                  │ allocs/op  │ allocs/op   vs base              │
IndexMarshal-16     49.00 ± 2%   49.50 ± 1%       ~ (p=1.000 n=8)
IndexUnmarshal-16   219.0 ± 0%   207.0 ± 0%  -5.48% (p=0.000 n=8)
geomean             103.6        101.2       -2.28%
```